File: ui/src/main/kotlin/listenwithrandos/state/Store.kt
package listenwithrandos.state

import kotlinext.js.js
import kotlinx.serialization.json.JSON
import redux.RAction
import redux.Store
import redux.WrapperAction
import redux.rEnhancer
import kotlin.browser.window

fun createStore(): Store<State, RAction, WrapperAction> {
    val storedState = window.localStorage.getItem("state")?.let {
        val state = JSON.parse<State>(it)
        console.log(state)
        state
    }

    val preloadedState = storedState ?: js {}

    val store = redux.createStore(
        reducer,
        preloadedState,
        rEnhancer()
    )

    store.subscribe {
        console.log(store.getState())
        val state = JSON.stringify(store.getState().unsafeCast<State>())
        println(state)
        window.localStorage.setItem("state", state)
    }

    return store
}

File: ui-new/src/domain/store/store.js
// @flow

import {applyMiddleware, createStore} from 'redux'
import type {Store} from 'redux'
import {reducer} from './reducer/reducer'
import type {State} from './reducer/reducer'
import type {AppAction} from './actions/actions'
import {createEpicMiddleware} from 'redux-observable'
import {epic} from './epic'

export type AppStore = Store<State, AppAction>

export const initializeStore = (): AppStore => {
  const storedState = window.localStorage.getItem('state')

  const preloadedState: $Shape<State> = storedState ? JSON.parse(storedState) : {}

  const epicMiddleware = createEpicMiddleware()

  const store = createStore(
    reducer,
    preloadedState,
    applyMiddleware(epicMiddleware),
  )

  epicMiddleware.run(epic)

  store.subscribe(() => {
    const state = store.getState()
    window.localStorage.setItem('state', JSON.stringify(state))
  })

  return store
}
